File: NodeDesignPatterns/command.js
"use strict";
let doc = `

                    +------------------+
                    |     Client       |
                    +------------------+
                             |
                             |
                             v

                    +------------------+
                    |     Command      |  <<interface>>
                    +------------------+
                    | + execute()      |
                    +------------------+
                             ^
                             |
          +------------------+------------------+
          |                                     |
          |                                     |
+----------------------+       +----------------------+
| ConcreteCommandA    |       | ConcreteCommandB     |
+----------------------+       +----------------------+
| - receiver          |       | - receiver           |
+----------------------+       +----------------------+
| + execute()         |       | + execute()          |
+----------------------+       +----------------------+
          |                                     |
          |                                     |
          v                                     v

+----------------------+       +----------------------+
|      ReceiverA       |       |      ReceiverB       |
+----------------------+       +----------------------+
| operationA()         |       | operationB()         |
+----------------------+       +----------------------+


                    +------------------+
                    |     Invoker      |
                    +------------------+
                    | + execute(cmd)   |
                    +------------------+
                             |
                             |
                             v
                         Command




`;
console.log(doc);
class Command {
}
class Light {
    lightOn() {
        console.log("light on");
    }
    lightOff() {
        console.log("light off");
    }
}
class Fan {
    start() {
        console.log("Fan on");
    }
    stop() {
        console.log("Fan off");
    }
}
class AC {
    coolOn() {
        console.log("AC on");
    }
    coolOff() {
        console.log("AC off");
    }
}
class LightOnCommand extends Command {
    light;
    constructor(light) {
        super();
        this.light = light;
    }
    execute() {
        this.light.lightOn();
    }
}
class LightOffCommand extends Command {
    light;
    constructor(light) {
        super();
        this.light = light;
    }
    execute() {
        this.light.lightOff();
    }
}
class ACOnCommand extends Command {
    ac;
    constructor(ac) {
        super();
        this.ac = ac;
    }
    execute() {
        this.ac.coolOn();
    }
}
class ACOffCommand extends Command {
    ac;
    constructor(ac) {
        super();
        this.ac = ac;
    }
    execute() {
        this.ac.coolOff();
    }
}
class FanOnCommand extends Command {
    fan;
    constructor(fan) {
        super();
        this.fan = fan;
    }
    execute() {
        this.fan.start();
    }
}
class FanOffCommand extends Command {
    fan;
    constructor(fan) {
        super();
        this.fan = fan;
    }
    execute() {
        this.fan.stop();
    }
}
// Client Code
const light = new Light();
const lightOn = new LightOnCommand(light);
lightOn.execute();

File: PythonDesignPatterns/command.py
"""
                    +------------------+
                    |     Client       |
                    +------------------+
                             |
                             |
                             v

                    +------------------+
                    |     Command      |  <<interface>>
                    +------------------+
                    | + execute()      |
                    +------------------+
                             ^
                             |
          +------------------+------------------+
          |                                     |
          |                                     |
+----------------------+       +----------------------+
| ConcreteCommandA    |       | ConcreteCommandB     |
+----------------------+       +----------------------+
| - receiver          |       | - receiver           |
+----------------------+       +----------------------+
| + execute()         |       | + execute()          |
+----------------------+       +----------------------+
          |                                     |
          |                                     |
          v                                     v

+----------------------+       +----------------------+
|      ReceiverA       |       |      ReceiverB       |
+----------------------+       +----------------------+
| operationA()         |       | operationB()         |
+----------------------+       +----------------------+


                    +------------------+
                    |     Invoker      |
                    +------------------+
                    | + execute(cmd)   |
                    +------------------+
                             |
                             |
                             v
                         Command


"""
print(__doc__)

from abc import ABC, abstractmethod

class Light:
    
    def light_on(self):
        print("light on")
    
    def light_off(self):
        print("light off")

class Fan:
    def start(self):
        print("Fan on")
    def stop(self):
        print("Fan off")


class AC:
    def cool_on(self):
        print("AC on")
    def cool_off(self):
        print("AC off")



class command(ABC):
    @abstractmethod
    def execute(self):
        pass



class LightOncommand(command):
    def __init__(self,light):
        self.light=light
    def execute(self):
        return self.light.light_on()

class LightOffcommand(command):
    def __init__(self,light):
        self.light=light
    def execute(self):
        return self.light.light_off()



class ACncommand(command):
    def __init__(self,ac):
        self.ac=ac
    def execute(self):
        return self.ac.cool_on()
class ACoffcommand(command):
    def __init__(self,ac):
        self.ac=ac
    def execute(self):
        return self.ac.cool_()
    


class FanOncommand(command):
    def __init__(self,fan):
        self.fan=fan
    def execute(self):
        return self.fan.start()

class FanOffcommand(command):
    def __init__(self,fan):
        self.fan=fan
    def execute(self):
        return self.fan.stop()
    



light=Light()
light_on=LightOncommand(light)
light_on.execute()
